### DISCUSSION TITLE
Better nonce handling with many/parallel transactions

### DISCUSSION THREAD

**Discussion by deanpress** [wevm/viem#1338] [Idea / Feature Request]
Currently, viem lacks built-in support for efficient nonce management when making many transactions from the same account in quick/parallel succession. This can lead to nonce collisions or inconsistency, resulting in transactions failing with invalid nonce errors.

I propose the addition of improved nonce management in viem,  specifically when it comes to many parallel transactions (similar to [ethers' NonceManager plug-in](https://github.com/ethers-io/ethers.js/blob/main/src.ts/providers/signer-noncemanager.ts)). Nonces should be managed automatically for each transaction sent from the same Ethereum address, ensuring that transactions are processed properly without being reliant on manual nonce implementations, and minimizing "nonce too low/high" errors.

When one out of many (semi) parallel transactions fails, new and pending transactions should still be broadcast/retried without nonce errors, even if no nonce was manually specified. 

Feature Details:

1. **Auto-increment Nonce**: The utility should keep track of the last used nonce for a given account and auto-increment it for the next transaction. Currently this appears to fail when many transactions are being broadcast in parallel.

2. **Concurrency Support**: Capability to handle multiple transactions in parallel, ensuring that each gets its own unique nonce.

3. **Error Handling**: Effective mechanisms to handle errors, including nonce collisions, to give a chance for transactions to be retried in viem's built-in retry system.

**Comment by joaofranciscoguarda** [comment]
That's a must need.

**Reply by salieflewis** [reply]
I wonder what this would look like in the context of viem. Do you think there is a way we could simply extend walletClient actions to facilitate this nonce tracking? Curious if you've put anymore thought into this since October.

cc: @deanpress @joaofranciscoguarda

**Reply by jxom** [reply]
We could probably integrate a nonce store inside Viem's [Local Accounts](https://viem.sh/docs/accounts/custom.html#usage). We can take a look at getting this in as part of the Viem v2 release.

**Reply by salieflewis** [reply]
That would be dope, happy to contribute to that as well!

**Comment by Akagi201** [comment]
I found this project implement a none manage. Anyone have read it? https://github.com/latticexyz/mud/blob/main/packages/common/src/createNonceManager.ts

**Reply by salieflewis** [reply]
I haven't come across that, but that is a great resource. Thanks for sharing!

**Comment by just-mitch** [comment]
For folks coming later, this was added in v2.15.0
https://viem.sh/docs/accounts/local/createNonceManager

**Comment by amilcarrey** [comment]
Here an implementation using @vercel/kv 

https://github.com/wevm/viem/discussions/2802
